streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -252,6 +288,111 @@ public void buildAndOptimizeTopology() {
         }
     }
 
+    private void maybePerformOptimizations(final Properties props) {
+
+        if (props != null && props.getProperty(StreamsConfig.TOPOLOGY_OPTIMIZATION).equals(StreamsConfig.OPTIMIZE)) {
+            LOG.debug("Optimizing the Kafka Streams graph for repartition nodes");
+            maybeOptimizeRepartitionOperations();
+        }
+    }
+
+    private void maybeOptimizeRepartitionOperations() {
+        StreamsGraphNode optimizedSingleRepartition = null;
+
+        for (final Map.Entry<StreamsGraphNode, Set<OptimizableRepartitionNode>> streamsGraphNodeSetEntry : keyChangingOperationsToOptimizableRepartitionNodes.entrySet()) {
+
+            StreamsGraphNode keyChangingNode = streamsGraphNodeSetEntry.getKey();
+
+            if (streamsGraphNodeSetEntry.getValue().isEmpty()) {
+                continue;
+            }
+
+            final StreamSourceNode parentSourceNode = (StreamSourceNode) findParentNodeMatching(keyChangingNode, n -> n instanceof StreamSourceNode);
+
+            if (parentSourceNode == null) {
+                throw new IllegalStateException(String.format("Can't find parent source node for %s ", keyChangingNode));
+            }
+
+            optimizedSingleRepartition = createRepartitionNode(keyChangingNode.nodeName(),
```

Comment:
Hmm.. why we can set the serdes for `optimizedSingleRepartition` as the `parentSourceNode`? These two nodes are ancestor and child of the `keyChangingNode`, which could change the key type right??

Reply:
Ah yes, bad assumption on my part.  For the moment I've changed the call to pass in `null` for the `Serde` objects and rely on what's specified in the configs so I could push the updates and get feedback on the changes.

With respect to the `Serde` usage, I guess this leaves us with a few alternatives.

1. Grab the `Serdes` from all the original repartition nodes and use those IF they are all the same type.
2. Update to allow for `Serde` inheritance by passing the key and value `Serde` instances and in turn pass those objects into the corresponding graph nodes.
3. Update the `KStream` interface to allow key changing methods to provide `Serialized` instances, although this is more heavyweight and requires a KIP.

WDYT?
\cc @mjsax  @vvcephei

Reply:
I think we cannot just pass in `null` and rely on default serde types from configs, since users are not expecting the repartition to happen here from the DSL, and they thought they have provided enough serde informations on places they "think" a repartition will happen. So observing a `ClassCastException` at a place they are not expected would be a bad experience.

As of now, without serde inheritance we may have to "restrict" the optimization, to only apply to cases where two or more repartition nodes are direct children of the common key-changing parent node. 

Moving forward we can consider the options @bbejeck provided, and personally I think serde inheritance would be a good place to start.

Reply:
I agree with Guozhang. If Serdes are different, it implies that we cannot apply the optimization (ie, it applies we would write different types into the repartition topic after the optimization -- falling back to default serdes does not resolve the issue, but we need to use two repartition topics -- one for each data type).